File: swift/11_sorts/Sorts.swift
import Foundation

/// 冒泡排序
///
/// - Parameter elements: 数组
/// - Returns: 返回值
public func bubbleSort<T>(_ elements: [T]) ->[T] where T: Comparable {
    var array = elements
    guard array.count > 1 else {
        return array
    }
    for i in 0..<array.count {
        // 提前退出标志位
        var flag = false
        for j in 0..<array.count - i - 1 {
            if array[j] > array[j+1] {
                array.swapAt(j+1, j)
                // 此次冒泡有数据交换
                flag = true
            }
        }
        if (!flag) {
            break
        }
    }
    return array
}


/// 插入排序
///
/// - Parameter elements: 数组
/// - Returns: 返回值
public func insertionSort<T>(_ elements: [T]) -> [T] where T: Comparable {
    var array = elements
    guard array.count > 1 else {
        return array
    }
    for i in 1..<array.count {
        let value = array[i]
        var j = i - 1;
        // 查找要插入的位置并移动数据
        for p in (0...j).reversed() {
            j = p
            if array[p] > value {
                array[p+1] = array[p]
            } else {
                break
            }
        }
        array[j+1] = value
    }
    return array
}


/// 选择排序
///
/// - Parameter elements: 数组
/// - Returns: 返回值
public func selectionSort<T>(_ elements: [T]) -> [T] where T: Comparable {
    var array = elements
    guard array.count > 1 else {
        return array
    }
    for i in 0..<array.count {
        // 查找最小值
        var minIndex = i
        var minValue = array[i]
        for j in i..<array.count {
            if array[j] < minValue {
                minValue = array[j]
                minIndex = j
            }
        }
        // 交换
        array.swapAt(i, minIndex)
    }
    return array
}

File: java/11_sorts/Sorts.java
package sorts;

/**
 * 冒泡排序、插入排序、选择排序
 *
 * Author: Zheng
 */
public class Sorts {

  // 冒泡排序，a是数组，n表示数组大小
  public static void bubbleSort(int[] a, int n) {
    if (n <= 1) return;

    for (int i = 0; i < n; ++i) {
      // 提前退出标志位
      boolean flag = false;
      for (int j = 0; j < n - i - 1; ++j) {
        if (a[j] > a[j+1]) { // 交换
          int tmp = a[j];
          a[j] = a[j+1];
          a[j+1] = tmp;
          // 此次冒泡有数据交换
          flag = true;
        }
      }
      if (!flag) break;  // 没有数据交换，提前退出
    }
  }

  // 插入排序，a表示数组，n表示数组大小
  public static void insertionSort(int[] a, int n) {
    if (n <= 1) return;

    for (int i = 1; i < n; ++i) {
      int value = a[i];
      int j = i - 1;
      // 查找要插入的位置并移动数据
      for (; j >= 0; --j) {
        if (a[j] > value) {
          a[j+1] = a[j];
        } else {
          break;
        }
      }
      a[j+1] = value;
    }
  }

  // 选择排序，a表示数组，n表示数组大小
  public static void selectionSort(int[] a, int n) {
    if (n <= 1) return;

    for (int i = 0; i < n - 1; ++i) {
      // 查找最小值
      int minIndex = i;
      for (int j = i + 1; j < n; ++j) {
        if (a[j] < a[minIndex]) {
          minIndex = j;
        }
      }
      
      // 交换
      int tmp = a[i];
      a[i] = a[minIndex];
      a[minIndex] = tmp;
    }
  }

}
